cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysisStage.qll

```diff
@@ -1062,6 +1062,20 @@ module RangeStage<DeltaSig D, BoundSig<D> Bounds, LangSig<D> LangParam, UtilSig<
         or
         upper = false and delta = D::fromFloat(-D::toFloat(d_max).abs() + 1)
       )
+      or
+      exists(
+        SemZeroBound bLeft, SemZeroBound bRight, D::Delta dLeft, D::Delta dRight, boolean fbeLeft,
+        boolean fbeRight, D::Delta odLeft, D::Delta odRight, SemReason rLeft, SemReason rRight
+      |
+        boundedMulOperand(e, upper, bLeft, true, dLeft, fbeLeft, odLeft, rLeft) and
+        boundedMulOperand(e, upper, bRight, false, dRight, fbeRight, odRight, rRight) and
+        delta = D::fromFloat(D::toFloat(dLeft) * D::toFloat(dRight)) and
+        fromBackEdge = fbeLeft.booleanOr(fbeRight)
+      |
+        b = bLeft and origdelta = odLeft and reason = rLeft
+        or
+        b = bRight and origdelta = odRight and reason = rRight
```

Comment:
Why do we match both `bLeft` and `bRight`, rather than just one of these - if they're both a zero bound (which I gather means `0 + delta`) and we've already computed the multiplied `delta` bound?

Reply:
Indeed. The zero bound should be a unit type, so there's no need for it to even be a column in `boundedMulOperand`. Just do `b instanceof SemZeroBound` here.

Reply:
Fixed in fe487a1ae80746b4181fd1d40e7fdae46db4995d.

Reply:
I think the remainder case could do with a similar fix.